### PR TITLE
Adds support for `jaccard_coefficient`

### PR DESCRIPTION
**Pull request by rlratzel**
Adds support for `jaccard_coefficient` to nx-cugraph.

This includes a test, but relies largely on the existing test coverage provided by NetworkX. The test included here could (should) be submitted to NetworkX though in a separate PR, since it is not covering anything unique to nx-cugraph.

A benchmark is also included, with results showing 2-4X speedup. I've seen much, much larger speedup on a different graph (large movie review bipartite graph, showing 966s for NX, 2s for nx-cugraph = ~500X), so I need to investigate further.  This investigation need not prevent this PR from being merged now though.

image

### CODE REVIEW

**Review by nv-rliu — APPROVED** [review]

**Review by nv-rliu — APPROVED** [review]
lgtm!

File: nx_cugraph/algorithms/link_prediction.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+@not_implemented_for("directed")
+@not_implemented_for("multigraph")
+@networkx_algorithm(version_added="25.02", _plc="jaccard_coefficients")
+def jaccard_coefficient(G, ebunch=None):
+    if ebunch is None:
+        # FIXME: is there a more efficient way to do this (on GPU or
+        # otherwise)?
+        ebunch = list(nx.non_edges(G))
+        if not ebunch:
+            return iter([])
```
Comment (eriknw) [inline nx_cugraph/algorithms/link_prediction.py:34]: Here's an alternative:
```suggestion
    G = _to_undirected_graph(G)
    if ebunch is None:
        A = cp.tri(G._N, G._N, dtype=bool)
        A[G.src_indices, G.dst_indices] = True
        src_indices, dst_indices = cp.nonzero(~A)
        if src_indices.size == 0:
            return iter([])
        src_indices = src_indices.astype(index_dtype)
        dst_indices = dst_indices.astype(index_dtype)
```
there are other variations and alternatives; for example, you could use a mask:
```python
        mask = G.src_indices < G.dst_indices
        A[G.src_indices[mask], G.dst_indices[mask]] = True
```
Observe that these go straight to indices, so there is no need to call e.g. `G._list_to_nodearray(src_indices)` to do remapping from keys to indices.

Also, if possible, I would advise against using the graph object (such as calling `nx.non_edges(G)`) before converting the graph via e.g. `G = _to_undirected_graph(G)`. I think it's best to make fewer assumptions about the input.

File: nx_cugraph/algorithms/link_prediction.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+    G = _to_undirected_graph(G)
+
+    # FIXME: this zip() call appears to be the performance bottleneck for this
+    # function. Is there a better way?
+    (u, v) = zip(*ebunch)
```
Comment (eriknw) [inline nx_cugraph/algorithms/link_prediction.py:40]: As noted above, this can be avoided when `ebunch` is not given. When `ebunch` is given, I don't think there is a better way.

File: nx_cugraph/algorithms/link_prediction.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+    try:
+        # Convert the ebunch lists to cupy arrays for passing to PLC, possibly
+        # mapping to integers if the Graph was renumbered.
+        # Allow the Graph renumber lookup (if renumbering was done) to check
+        # for invalid node IDs in ebunch.
+        u = G._list_to_nodearray(u)
+        v = G._list_to_nodearray(v)
+    except KeyError as n:
+        raise nx.NodeNotFound(f"Node {n} not in G.")
```
Comment (eriknw) [inline nx_cugraph/algorithms/link_prediction.py:50]: It's cute that the `str(n)` of the `KeyError` is simply the key so that this works :+1:

File: nx_cugraph/algorithms/link_prediction.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    else:
```
Comment (eriknw) [inline nx_cugraph/algorithms/link_prediction.py:51]: Having an `else:` block here is unnecessary. I would delete this `else:` to simplify the control flow and remove one level of indentation.

File: nx_cugraph/algorithms/link_prediction.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+                    [
+                        G.src_indices,
+                        G.dst_indices,
+                        G._node_ids if G._node_ids is not None else cp.ndarray(0),
+                    ]
+                )
+            )
+            invalid = cp.setdiff1d(ebunch_nodes, graph_nodes, assume_unique=True)
+            if len(invalid) > 0:
+                raise nx.NodeNotFound(f"Node {invalid.tolist()[0]} not in G.")
```
Comment (eriknw) [inline nx_cugraph/algorithms/link_prediction.py:72]: Thanks for being careful about behavior with missing nodes. There are no helper functions for this yet; each algorithm behaves differently and does it themselves. I think it probably makes sense to have a helper function.

Since we have the indices, we just need to know if any are out of bounds, so it's sufficient to check bounds with min/max.

Continuing from my `suggestion` code above (these aren't meant to be merged as-is), this could be:
```suggestion
    else:
        (u, v) = zip(*ebunch)
        try:
            src_indices = G._list_to_nodearray(u)
            dst_indices = G._list_to_nodearray(v)
        except KeyError as n:
            raise nx.NodeNotFound(f"Node {n} not in G.")
        if G.key_to_id is None and (
            (n := src_indices.max()) >= G._N
            or (n := dst_indices.max()) >= G._N
            or (n := src_indices.min()) < 0
            or (n := dst_indices.min()) < 0
        ):
            raise nx.NodeNotFound(f"Node {n} not in G.")  
```
Is `NodeNotFound` tested? If not, can we add a test to nx-cugraph or upstream?

File: nx_cugraph/algorithms/link_prediction.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+    # Note that Jaccard similarity must run on a symmetric graph.
+    # FIXME: PLC will symmetrize the graph if told to, but the symmetrize flag
+    # to _get_plc_graph() appears to symmetrize using cupy and does other
+    # things (cast to 64bit, etc.). Can we let PLC do the symmetrization if the
+    # symmetrize flag is set instead?
```
Comment (eriknw) [inline nx_cugraph/algorithms/link_prediction.py:78]: > Can we let PLC do the symmetrization if the symmetrize flag is set instead?

Quite possibly. It should be straightforward to experiment now since this ought to be covered by tests.

I don't think this comment belongs here. It reads as either a comment that belongs in a PR discussion, or if we want a bread crumb for us to find later in the code, it should be findable in `_get_plc_graph`. Can you clean up or tighten some of your comments so that they reflect what is most helpful to future readers and perhaps add some comments to PR?

File: nx_cugraph/algorithms/link_prediction.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    (u, v, p) = plc.jaccard_coefficients(
+        resource_handle=plc.ResourceHandle(),
+        graph=G._get_plc_graph(symmetrize=None),
```
Comment (eriknw) [inline nx_cugraph/algorithms/link_prediction.py:81]: `symmetrize=None` is not required. It is the default. Undirected graphs are always symmetric anyway.
```suggestion
        graph=G._get_plc_graph(),
```

File: nx_cugraph/algorithms/link_prediction.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    p = p.tolist()
+
+    # zip() returns a zip object, which is different than default NX but should
+    # still be valid since it's an iterator of 3-tuples.
+    return zip(u, v, p)
+
+
+# @jaccard_coefficient._can_run is always True (default)
+
+# @jaccard_coefficient._should_run is always True (default)
```
Comment (eriknw) [inline nx_cugraph/algorithms/link_prediction.py:99]: Can we delete these comments? Would you want to add similar comments for _every_ function? It's fine to remind us reviewers that we always can and should run, but keeping these comments in the code seems unnecessary and noisy to me.

File: nx_cugraph/algorithms/link_prediction.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        do_expensive_check=False,
+    )
+
+    u = G._nodearray_to_list(u)
+    v = G._nodearray_to_list(v)
+    p = p.tolist()
+
+    # zip() returns a zip object, which is different than default NX but should
+    # still be valid since it's an iterator of 3-tuples.
+    return zip(u, v, p)
```
Comment (eriknw) [inline nx_cugraph/algorithms/link_prediction.py:94]: As discussed with the core NX team, NetworkX does not differentiate between generators (functions that yield) and functions that return iterators. They switch between the two interchangeably and they do not update docstring or do a deprecation cycle.

Hence, nx-cugraph and networkx both return an iterator of 3-tuples.

If we really wanted to, we could do e.g. `yield from zip(u, v, p)`.

**Review by eriknw — CHANGES_REQUESTED** [review]

File: nx_cugraph/algorithms/link_prediction.py (hunk 1 above)
Comment (rlratzel) [reply nx_cugraph/algorithms/link_prediction.py:34]: ok thanks, I'm going to benchmark and apply one of these suggestions (if they're faster, which I'm assuming they are).

I'm a bit confused by this statement though:
> I would advise against using the graph object (such as calling nx.non_edges(G)) before converting the graph via e.g. G = _to_undirected_graph(G). I think it's best to make fewer assumptions about the input.

If I need to call a NX function like `nx.non_edges()`, why would I first convert the input to a `CudaGraph` using `_to_undirected_graph()`?  Shouldn't we be assuming the input is a compatible NX `Graph` type, or are we allowing `CudaGraph` objects to be passed as well?  I suspect I'm missing something. (side note: this is ideal info for a docstring, but I believe the decorator bubbles them up to the corresponding nx function docstring, which I wouldn't want here.)

File: nx_cugraph/algorithms/link_prediction.py (hunk 4 above)
Comment (rlratzel) [reply nx_cugraph/algorithms/link_prediction.py:51]: Ah, yes, I vaguely remember thinking I needed that, possibly from a prior implementation that worked differently, but I'd also prefer to not have it.

File: nx_cugraph/algorithms/link_prediction.py (hunk 5 above)
Comment (rlratzel) [reply nx_cugraph/algorithms/link_prediction.py:72]: This LGTM.

> Is NodeNotFound tested? If not, can we add a test to nx-cugraph or upstream?

Yes, the NX tests have a few tests to ensure it's raised properly ([here](https://github.com/networkx/networkx/blob/f2c43bc6d22f470497ca2bd97c6458969ecf8320/networkx/algorithms/tests/test_link_prediction.py#L43)'s one) and they helped me discover a few edge cases IIRC.

File: nx_cugraph/algorithms/link_prediction.py (hunk 7 above)
Comment (rlratzel) [reply nx_cugraph/algorithms/link_prediction.py:81]: This is here as a companion to the FIXME above, but I can remove it since I'm also moving the FIXME to `_get_plc_graph()`.

**Review by eriknw — APPROVED** [review]
LGTM! Thanks for the updates. I left one minor suggestion.

Did you ever dig into understanding the benchmark performance? What's the procedure for updating the benchmark docs?

Adding jaccard makes me want to also add `adamic_adar_index` and `preferential_attachment` (should be easy to implement, and they show up in examples and learning material, but they're not commonly used by networkx dependents--but neither is jaccard!).

File: nx_cugraph/algorithms/link_prediction.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+            u_indices = G._list_to_nodearray(u)
+            v_indices = G._list_to_nodearray(v)
+        except KeyError as n:
+            raise nx.NodeNotFound(f"Node {n} not in G.")
+
+        # If G was not renumbered, then the ebunch nodes must be explicitly
+        # checked. If not done, plc.jaccard_coefficients() will accept node IDs
+        # not in the graph and return a coefficient of 0 for them, which is not
+        # compatible with NX.
+        if (not hasattr(G, "key_to_id") or G.key_to_id is None) and (
```
[the remark this answers is not in the document]
Comment (rlratzel) [reply nx_cugraph/algorithms/link_prediction.py]: I ended up refactoring the `ebunch` node check in order to pass some tests added to ensure we behave like NX. The change consolidates the additional valid node checks in `_list_to_nodearray`, but let me know if you see any issues.

**Comment by rlratzel** [comment]
> Did you ever dig into understanding the benchmark performance? What's the procedure for updating the benchmark docs?

IIRC, nx-cugraph speedup over NX increases as average degree and the size of `ebunch` increases, and graph size doesn't come into play since Jaccard is just comparing arbitrary pairs of nodes and their neighbors.  We could add an additional dataset (or generate one) to emphasize this in our benchmarks.

> What's the procedure for updating the benchmark docs?

@nv-rliu are there docs (README, something else) on updating the benchmark docs?

**Comment by nv-rliu** [comment]
> @nv-rliu are there docs (README, something else) on updating the benchmark docs?

There aren't any new docs on our side for updating the benchmark numbers in the table. The process would be to update `bench_algos.py` to contain the benchmark we want to include, and then repeat the same process with Perflab and specify that we want results for the new algo (i.e. Jaccard) and then once we get those we can include them in the table. Once this is merged, I can get right on that.

File: nx_cugraph/classes/graph.py [hunk 11]
```diff
@@ -1152,6 +1152,10 @@ def _nodearray_to_list(self, node_ids: cp.ndarray[IndexValue]) -> list[NodeKey]:
     def _list_to_nodearray(self, nodes: list[NodeKey]) -> cp.ndarray[IndexValue]:
         if (key_to_id := self.key_to_id) is not None:
             nodes = [key_to_id[node] for node in nodes]
+        else:
+            valids = [isinstance(n, int) and n >= 0 and n < self._N for n in nodes]
+            if not all(valids):
+                raise ValueError(nodes[valids.index(False)])
```
Comment (eriknw) [inline nx_cugraph/classes/graph.py:1158]: Heh, right, values like 4.5 are also invalid.

There are many types of integers floating around so `isinstance(n, int)` may be inadequate, and `isinstance(n, numbers.Integral)` is slow, so here's an alternative
```python
N = self._N
for node in nodes:
    try:
        node = index(node)  # Ensure integral value
    except TypeError:
        raise KeyError(node) from None
    if node < 0 or node >= N:
        raise KeyError(node)
```
where we import `index` from `operators`.

This is a little more strict than NetworkX
```python
In [2]: G = nx.complete_graph(10)

In [3]: 1 in G
Out[3]: True

In [4]: 1.0 in G
Out[4]: True
```
so another alternative could be
```python
N = self._N
for node in nodes:
    try:
        n = int(node)
    except TypeError:
        raise KeyError(node) from None
    if n != node or n < 0 or n >= N:
        raise KeyError(node)
```

Also, what do you think of making this verification opt-in by adding a keyword argument to the method? I'm conflicted, b/c I like performance. We have sometimes played a little fast and loose in the name of performance, and we may not catch every invalid node that NetworkX would catch. This has been somewhat intentional: functions in networkx don't have consistent behavior, so I would want to have e.g. randomized tests like Ross talked about to stress both networkx and networkx backends for exceptional behavior. OTOH, it's probably safest to do this check here.

**Comment by eriknw** [comment]
Maybe a silly question, but can `plc.all_pairs_jaccard_coefficients` be useful here? How is it different from `plc.jaccard_coefficients`?

File: nx_cugraph/classes/graph.py (hunk 11 above)
Comment (rlratzel) [reply nx_cugraph/classes/graph.py]: > Also, what do you think of making this verification opt-in by adding a keyword argument to the method?

I like that idea. It seems like many of our functions could take an argument that essentially says "skip input verification because I know they're valid" for use cases optimized for performance. This seems like a good topic for the dispatching meeting, even though what you're proposing is or could be specific to nx-cugraph's kwargs.

Comment (rlratzel) [reply nx_cugraph/classes/graph.py]: I added this feature request for the opt-in suggestion above: https://github.com/rapidsai/nx-cugraph/issues/70

**Comment by rlratzel** [comment]
> Maybe a silly question, but can `plc.all_pairs_jaccard_coefficients` be useful here? How is it different from `plc.jaccard_coefficients`?

@jnke2016 can answer in more detail, but the `all_pairs` functions in PLC and cugraph were added to support various use cases where 1) users wanted to default the ebunch list (or equivalent param based on the function) in a specific way so they didn't need to build it themselves. In this case, the default is the 2-hop neighbors for every node, which is unfortunately different than the NX default of all non-edges. And 2) it adds a parameter (`top_k`) to return only the top `k` values of the total results to save memory.  Since we need to default to non-edges and NX doesn't support the notion of `top_k` (that I'm aware of), we unfortunately wouldn't see any benefit.

File: nx_cugraph/classes/graph.py (hunk 11 above)
Comment (eriknw) [reply nx_cugraph/classes/graph.py]: My original thought was to keep this an internal option--i.e., keyword--to `_list_to_nodearray` (and maybe others) to control whether to do this check. E.g., some networkx algorithms may or may not raise.
